### ISSUE TITLE
Simplify customisation that is used to implement strings

### ISSUE THREAD

**Issue by agoose77** [scikit-hep/awkward#2432]
### Description of new feature

There has been a long-standing awareness that our string abstraction is both a special case (privileged in our codebase) and also something that users might want to customise. Consequently, we have mechanisms like custom broadcasting that users can hook into to implement the same features.

Now that Awkward is more mature, we can more conclusively say that few/no users are actually making use of these mechanisms (@jpivarski might have data on this, or I can grep it!)
However, this might be a chicken-and-egg problem; the extension system is not sufficiently mature or comprehensive that users could implement strings without Awkward's special casing.

Specifically, we have known bugs that are trickier to solve for our extension mechanisms, such as string-string broadcasting. 

There are clear use cases for extending strings, e.g. IP addresses:
- Users want IP addresses to broadcast like strings (string-like)
- Users want IP addresses to have IP address methods (IP address-like)

Now, IP addresses could be implemented as tuples (@jpivarski could you expand on the reason not to do this?), but let's consider that we want to implement them as strings. Really, as @jpivarski pointed out in a meeting today, strings are a special example of binary data (an escape hatch, of sorts). 

It would appear that the best path forward is to simplify our extension mechanisms to make string-like custom types _easier_, whilst removing the kinds of dials that users can turn.
Namely, we could introduce a string-like table to `ak.behavior` that enumerates the string-like type names (that are used with `__array__`). This enables other types to gain string-like broadcasting, without opening the door to esoteric broadcasting rules. Note that to gain string-like _methods_, the names placed in this table should correspond to array classes that _inherit_ from the string behaviour classes. Note that, only some of Awkward's logic is implemented in behaviour classes — much is implemented inside the Awkward code base.

Specifically, string-like names imply the following 2D list containing a 1D `NumpyArray` of `uint8`. Concretely the "string-like" typeclass (borrowed from Haskell) means that the list of list of `uint8` should be treated as a list of atoms. By applying such a restrictive definition, we can avoid ambiguity when handling all of the string-related Awkward rules.

This relates to another conversion on type _inheritance_. That will be discussed separately.

## Related Issues
- #1682

**Comment by jpivarski** [comment]
**IP addresses as strings:** they usually arrive as strings. In fact, as bytestrings—not `127.0.0.1`, but `b'\x7f\x00\x00\x01'`. Libraries of algorithms on them will likely expect that, too. It would be a pain to convert them into tuples and back again.

---------------------------

**What are all the ways that people use `ak.behavior`?**

  1. https://pivarski-princeton.s3.amazonaws.com/GitHub-CMSSW-user-nonfork-raw-data-1Mcut.tar (82.7 GB)
  2. https://pivarski-princeton.s3.amazonaws.com/GitHub-ROOT-user-nonfork-raw-data-1Mcut.tar (44.3 GB)

are the full contents of all the non-fork repos of users who have either (1) forked CMSSW before Feb 14, 2023 or (2) had any GitHub interaction at all with ROOT between 2017 and April 13, 2013. After untarring them (in a new, empty directory), they're thousands of username directories containing tens of thousands of zipped-tarball repos.

I'm running this on all of them:

```python
import glob
import tarfile
import re

for filename in glob.glob("/home/jpivarski/storage/data/GitHub-*-raw-data-1Mcut/*/*.tgz"):
    try:
        with tarfile.open(filename) as repo_tgz:
            for info in repo_tgz:
                try:
                    if info.name.lower().endswith(".py"):
                        with repo_tgz.extractfile(info) as file:
                            text = file.read()
                            for match in re.finditer(b"^\s*(.*\.behavior\s*\[.*)$", text, re.M):
                                try:
                                    print(
                                        info.name + "\t" + match.group(1).decode(errors="surrogateescape"),
                                        flush=True,
                                    )
                                except:
                                    pass
                except:
                    pass
    except:
        pass

```

I'll post the results here. The `info.name` will contain the `USERNAME/REPONAME/full/path/to/file`.

For more information, see the README on https://github.com/jpivarski-talks/2023-05-09-chep23-analysis-of-physicists.

**Comment by jpivarski** [comment]
Wow, that's not many results at all:

```
ssrothman/EECAnalysis/processing/EECProcessor.py	fname = data.behavior["__events_factory__"]._partition_key.replace("/", "-")
PKUfudawei/NanoAnalysis/src/processors/Processor.py	tokens = self.event.behavior["__events_factory__"]._partition_key.split('/')
maxgalli/FlashSimStudies/preprocessing/extract_photons.py	f_out = events.behavior["__events_factory__"]._partition_key.replace(
nikoladze/physlite_experiments/physlite_experiments/behavior.py	return self.behavior["__events__"][0]()
```

  * [ssrothman/EECAnalysis/processing/EECProcessor.py](https://github.com/ssrothman/EECAnalysis/blob/1b9e77caa823e3dc485715182605ea223f8b44ea/processing/EECProcessor.py#L445)
  * [PKUfudawei/NanoAnalysis/src/processors/Processor.py](https://github.com/PKUfudawei/NanoAnalysis/blob/e87a17fdf8428f022f61c90646f96c3f54173620/src/processors/Processor.py#L247)
  * [maxgalli/FlashSimStudies/preprocessing/extract_photons.py](https://github.com/maxgalli/FlashSimStudies/blob/74b01cb80e9ce4141014c5b0bb65dff55db64ebc/preprocessing/extract_photons.py#L159-L161)
  * [nikoladze/physlite_experiments/physlite_experiments/behavior.py](https://github.com/nikoladze/physlite_experiments/blob/e640bb3b906f5da037ba60e1a143d5f44347c638/physlite_experiments/behavior.py#L11)

There were some instances of `ak.behavior.update(some_dict)` but that dict came from some library (not included in the personal, non-fork repos). There were also uses of `@ak.mixin_class`, which sets behaviors, but only for records and arrays of records.

I think it's safe to say that _nobody_ is overloading `__array__` parameters.

**Comment by agoose77** [comment]
Huh, interesting! I would have predicted more than that. I wonder how these results change for local behaviours in `array.behavior`? These would likely be found with a non-delimited regex like `r"\bbehavior\["`, module some false positives.

**Comment by agoose77** [comment]
@jpivarski and I discussed this today. We concluded that there are two favoured options to solving the problem of custom strings with string-like broadcasting:
1. Arrays with `__array__: "string"` or `__array__: "bytestring"`, **AND** an optional `__string__` "flavour" parameter.
2. Arrays with `__array__: "string"` or `__array__: "bytestring"`, **OR** with a `__string__` "flavour" parameter.

Before comparing these approaches, it's worth thinking about what we want from custom strings:
- String-like broadcasting
- String-like typestrings (already supported via `__typestr__`)
- Well-defined rules for mixing string types
- Custom behavior classes (e.g. ip-address methods)

Jim also pointed out that we might not need the behavior classes by default; the getitem logic already exists in our content objects. 

Option (1) does not admit distinct array classes _unless_ we extend the arrayclass lookup mechanism to consider both `__array__` and `__string__`. The easiest route would be:
1. Is the array a `__array__: "string"` (or bytestring)?
2. Does the array have a `__string__` variant?
   1. Does the `__string__` name exist in `ak.behavior`? → Use this class
4. Does the `__array__` name exist in `ak.behavior` (yes, always)? → Use this class

The problem with the above route is that the nominal type is now `__array__ OR __string__`. This dilutes the meaning of `__array__`, and makes it redundant if `__string__` is set.
Or, we state that the nominal type of a custom string is `(__array__, __string__)`, e.g. `("string", "ip-address")`. This means custom strings arrayclasses would be registered with `ak.behavior["string", "ip-address"]` or `ak.behavior["bytestring", "ip-address"]`. This feels unbalanced to me.

Option (2) preserves `__array__` as the _nominal_ type. There remains a 1:1 relationship between the array class, type string, etc, and `__array__`. However, there is no longer a way to express "this is a custom _byte_-string". Therefore, we probably want `__bytestring__` as a complementary parameter that says "this is a custom bytestring".

Crucially, option (2) allows us to decouple the _nominal_ type from the "dtype" (even though strings are not a dtype in awkward). 

I'm biased in favour of (2), but I'd like to discuss this further.

**Comment by jpivarski** [comment]
I like option (2) as well.

To be specific, these are the classes that might not be needed or even used anymore (possible dead code):

https://github.com/scikit-hep/awkward/blob/ca7637a0b4fa0aba0337095a604145e2491df26f/src/awkward/behaviors/string.py#L12-L107

When extracting a single string from an `ak.Array`, we don't return an instance of `StringBehavior` (as we would with ordinary overloads); we return a Python `str`. (Similarly for `bytes`.) This was raised as an issue in #873 and then spent some time as a draft PR in #874. Nobody complained, so it was implemented.

In fact, the following searches don't turn up any references to (not definitions of) the classes:

  * [repo:scikit-hep/awkward StringBehavior](https://github.com/search?q=repo%3Ascikit-hep%2Fawkward+StringBehavior&type=code)
  * [repo:scikit-hep/awkward ByteStringBehavior](https://github.com/search?q=repo%3Ascikit-hep%2Fawkward+ByteStringBehavior&type=code)
  * [repo:scikit-hep/awkward CharBehavior](https://github.com/search?q=repo%3Ascikit-hep%2Fawkward+CharBehavior&type=code)
  * [repo:scikit-hep/awkward ByteBehavior](https://github.com/search?q=repo%3Ascikit-hep%2Fawkward+ByteBehavior&type=code)

(apart from tests and documentation). I highly doubt that anyone uses these objects outside our codebase, since they don't link to anything in the codebase, so they must be dead code.

It would also be fairly easy to deprecate them, now. Just add a warning to the `__init__` and then `super().__init__(*args, **kwargs)`.

**Comment by agoose77** [comment]
On that note - we don't use all of the behavior class methods for strings/chars, but we use some:
- `ByteBehavior.__str__`, `.__repr__`
- `CharBehavior.__str__`, `__repr__`

I wonder whether custom `__str__` / `__repr__` should _also_ be assigned using `ak.behavior`, so that nested types can have nice reprs without having to resolve the array class. Then, I think, we can drop the need for a custom class in all cases, provided that we fix (if needed) the `__cast__` and scalar promotion of Python strings for `__eq__`, etc.

**Comment by agoose77** [comment]
One of the challenges here is the coupling between behavior classes (via `__array__` and `__record__` parameters) and array types (`__array__ = "string"`, ...). The intention that string classes be user-customisable _without_ requiring a significant rework of how far behaviors propagate into the codebase is one motivation, but another is generally figuring out how we _think_ about the `__array__` parameters vs other type parameters.

I think it would be helpful to start with describing some use cases:

1. Implementing a custom behavior for a categorical array
2. Implementing a custom behavior for a string, e.g. IP address
3. Implementing a custom behavior for an array with units

To compare "types" between two arrays, we have the following:
- `NumpyArray` — do both arrays have the same primitive?
- `RecordArray` — do both arrays have the same `__record__` **and** are their structures type-comparable?
- `ListOffsetArray`, ..., — do both arrays have the same `__array__`, and are their contents type-comparable?

It's clear to me that `__array__` and `__record__` are quite different, beyond their association with different content classes. Unlike `__record__`, `__array__` is used to implement non-behavior customisation that precludes the use of custom behaviors. 

I propose we introduce a new `__kind__` parameter that can currently be one of `("string", "categorical")`. We can introduce constraints upon which contents can be assigned with which values of `__kind__`. Introducing this parameter would allow us to separate user-provided names from built-in names, permitting custom strings and categoricals, e.g.
```python
my_string = ak.contents.ListOffsetArray(
    ak.index.Index64([0, 3, ..., 9]),
    ak.contents.NumpyArray(
        np.array(..., dtype=np.uint8)
    ),
    parameters={
        "__kind__": "string",
        "__array__": "ip-address"
    }
)
```

The nominal type precedence is then:
- `__array__`
- `__kind__`

and is used to resolve behaviors. Meanwhile, string-specific features check `__kind__`, or fall back upon `__array__` for legacy strings.

In this new formulation, we have the following interpretation of each parameter:
- `__array__` — the nominal type of a list (OR, the kind of a "legacy" string or categorical)
- `__record__` — the nominal type of a record
- `__kind__` — the built-in nominal type of a list (e.g. `string` or `categorical`)

Built-in types (strings, categoricals) look at `__kind__`, but high-level features like behaviors look at the resolved nominal type, i.e. via the precedence above.

Units also fit into this. Instead of implementing units with behaviors, I think we should implement them with a special `__units__` parameter, as they are orthogonal to the nominal type system.

**Comment by jpivarski** [comment]
I don't want to keep `ByteBehavior`/`CharBehavior` for the sake of their `__str__`/`__repr__`, but it is a good point that some overloaded types will want to overload how they appear in a string representation. Specifically, for this to catch it:

https://github.com/scikit-hep/awkward/blob/47975bdc01e514e49340e784bfb46278a44ebb94/src/awkward/_prettyprint.py#L70-L90

Records can easily do this. For array-like overloads (e.g. IP addresses), there's still a class to overload, right? If there's a class to overload, there's nothing special to do: users can define a `__str__`/`__repr__` on that class.

**Comment by agoose77** [comment]
I had the thought in #2528 that it's reasonable to keep the byte/char behaviors. We're considering making our _string_ abstraction built-in & customisable, but the byte/char arrays are not likely to be significantly impacted by this; custom strings use chars, and custom binary strings use bytes.

**Comment by agoose77** [comment]
Closed by #2549!
